examples-gallery/plot_ball_rolling_on_spinning_disc.py

```diff
@@ -0,0 +1,620 @@
+
+**Constants**
+
+- :math:`m_b`: mass of the ball [kg]
+- r : radius of the ball [m]
+- :math:`m_o` : mass of the observer [kg]
+- :math:`\\Omega`: final rotational speed of the disc
+  around the vertical axis [rad/sec]
+- :math:`\\alpha`: measure of the acceleration of the disc [1/sec]
+
+**States**
+
+- :math:`q_1, q_2, q_3`: generalized coordinates of the ball
+    w.r.t the disc [rad]
+- :math:`u_1, u_2, u_3`: generalized angual velocities of the ball [rad/sec]
+
+**Specifieds**
+
+- :math:`t_1, t_2, t_3`: Torque applied to the ball [N * m]
+
+**Additional parameters**
+
+- ``N``: inertial frame
+- ``A1``: frame fixed to the ball
+- ``A2``: frame fixed to the disc
+- ``O``: point fixed in N
+- ``CP``: contact point of ball with disc
+- ``Dmc``: center of the ball
+- :math:`m_{Dmc}`: position of observer
+- ``h``: intervall which opty should minimize.
+
+A video similar to this one, which I saw in something JM published
+gave me the idea.
+https://www.youtube.com/watch?v=3oM7hX3UUEU
+
+"""
+import sympy as sm
+import sympy.physics.mechanics as me
+import numpy as np
+from opty.direct_collocation import Problem
+from opty.utils import parse_free
+from scipy.interpolate import CubicSpline
+from scipy.optimize import root
+import matplotlib.pyplot as plt
+from matplotlib.animation import FuncAnimation
+from matplotlib import patches
+
+# %%
+# Initialize the variables.
+t = me.dynamicsymbols._t
+q1, q2, q3 = me.dynamicsymbols('q1 q2 q3')
+u1, u2, u3 = me.dynamicsymbols('u1 u2 u3')
+x, y, ux, uy = me.dynamicsymbols('x, y, ux, uy')
+t1, t2, t3 = me.dynamicsymbols('t1 t2 t3')
+
+# needed to mimick the time inm opty.
+T = sm.symbols('T', cls=sm.Function)
+Tdot, Tdotdot = sm.symbols('Tdot Tdotdot')
+
+mb, mo, g, r = sm.symbols('mb, mo, g, r')
+Omega, alpha = sm.symbols('Omega, alpha')
+
+N, A1, A2 = sm.symbols('N, A1 A2', cls=me.ReferenceFrame)
+O, CP, Dmc, m_Dmc = sm.symbols('O, CP, Dmc, m_Dmc', cls=me.Point)
+O.set_vel(N, 0)
+
+# %%
+# Determination of the holonomic constraints.
+#
+# If the ball rotates around the x axis by an angle q the
+# contact point will be move by -q1 * r in the A2.y direction.
+# If the ball rotates around the y axis by an angle q2 the
+# contact point will be move by q2 * r in the A2.x direction.
+# Hence the configuration constraints are:
+#
+# - x = r * q2
+# - y = -r * q1
+#
+# So the resulting speed constraints are:
+#
+# - d/dt(x) = r * d/dt(q2)
+# - d/dt(y) = -r * d/dt(q1)
+#
+#
+# The time t appears explicitly in the EOMs.
+# I copy the trick from JM in the example1.2.3 Parameter identification:
+# declare a function T(t), and then
+# pass the time as known trajectory. My OEMs also contain
+# d/dt(T(t)) and d^2/dx^2(T(t))
+# As T(t) = const * t I set these derivatives accordingly.
+
+# %%
+# Set up the equations of motion.
+# I do not use qdisc = sm.integrate(udisc, t), as this gives
+# a sm.Piecewise(..) result, likely not differentiable,
+# but use the result of this integration for alpha != 0.
```

Comment:
If you make appropriate assumptions on your variables you may be able to avoid a Piecewise.